File: packages/python/tests/integration/binding/repeatable_config_test.py
"""Binding-tier tests (real core, real fs) for repeatable ``config=`` (#588).

``DirSQL(config=...)`` accepts a list of ``.dirsql.toml`` paths that merge in
order (matching the Rust builder's repeatable ``.config()`` and the CLI's
repeatable ``-c``); a single string stays byte-identical to before.
"""

import os
import tempfile

import pytest

from dirsql import DirSQL


@pytest.fixture
def config_dir():
    with tempfile.TemporaryDirectory() as d:
        yield d


def _write(path, content):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        f.write(content)


def _table_config(name, glob):
    return f'[[table]]\nddl = "CREATE TABLE {name} (basename TEXT)"\nglob = "{glob}"\n'


def describe_repeatable_config():
    @pytest.mark.asyncio
    async def it_merges_tables_from_multiple_config_files(config_dir):
        _write(os.path.join(config_dir, "a.json"), "{}")
        cfg_a = os.path.join(config_dir, "a.dirsql.toml")
        cfg_b = os.path.join(config_dir, "b.dirsql.toml")
        _write(cfg_a, _table_config("alpha", "*.json"))
        _write(cfg_b, _table_config("beta", "*.json"))

        db = DirSQL(root=config_dir, config=[cfg_a, cfg_b])
        await db.ready()

        for table in ("alpha", "beta"):
            rows = await db.query(f"SELECT basename FROM {table}")
            assert [r["basename"] for r in rows] == ["a.json"], (
                f"table {table} from an accumulated config must be queryable"
            )

    @pytest.mark.asyncio
    async def it_accepts_a_single_config_string_unchanged(config_dir):
        _write(os.path.join(config_dir, "a.json"), "{}")
        cfg = os.path.join(config_dir, ".dirsql.toml")
        _write(cfg, _table_config("alpha", "*.json"))

        db = DirSQL(root=config_dir, config=cfg)
        await db.ready()
        rows = await db.query("SELECT basename FROM alpha")
        assert [r["basename"] for r in rows] == ["a.json"]

File: packages/ts/tests/integration/binding/repeatable-config.test.ts
// Binding-tier tests (real napi core, real fs) for repeatable `config` (#589).
//
// `new DirSQL({ config })` accepts a single path or an array of paths that
// merge in order (matching the Rust builder's repeatable `.config()` and the
// CLI's repeatable `-c`); a single string / `new DirSQL(path)` is unchanged.

import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

// The `on-file` hook emits `basename` itself rather than relying on the core
// injecting it, so the fixture stays green once stat-fact injection is removed;
// hook output overrides injection, keeping results identical meanwhile.
const basenameHook = `on-file = '''sh -c 'printf "[{\\"basename\\":\\"%s\\"}]" "\${1##*/}"' sh {path}'''`;

function tableConfig(name: string, glob: string): string {
  return `[[table]]\nname = "${name}"\nddl = "CREATE TABLE ${name} (basename TEXT)"\nglob = "${glob}"\n${basenameHook}\n`;
}

describe("repeatable config", () => {
  let dir: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-repeatcfg-"));
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it("merges tables from multiple config files", async () => {
    await writeFile(join(dir, "a.json"), "{}");
    const cfgA = join(dir, "a.dirsql.toml");
    const cfgB = join(dir, "b.dirsql.toml");
    await writeFile(cfgA, tableConfig("alpha", "*.json"));
    await writeFile(cfgB, tableConfig("beta", "*.json"));

    const db = new DirSQL({ root: dir, config: [cfgA, cfgB] });
    for (const table of ["alpha", "beta"]) {
      const rows = await db.query(`SELECT basename FROM ${table}`);
      expect(rows.map((r) => r.basename)).toEqual(["a.json"]);
    }
  });

  it("accepts a single config string unchanged", async () => {
    await writeFile(join(dir, "a.json"), "{}");
    const cfg = join(dir, ".dirsql.toml");
    await writeFile(cfg, tableConfig("alpha", "*.json"));

    const db = new DirSQL({ root: dir, config: cfg });
    const rows = await db.query("SELECT basename FROM alpha");
    expect(rows.map((r) => r.basename)).toEqual(["a.json"]);
  });
});
